File: demo/src/formatter.js
// A Formatter turns a report's title and rows into a string.
export class TextFormatter {
  format(title, rows) {
    const lines = [title, '='.repeat(title.length)];
    for (const row of rows) lines.push(`${row.label}: ${row.value}`);
    return lines.join('\n');
  }
}

// Registry mapping a format name to a formatter instance. New output
// formats register here; callers look one up by name. Today only 'text'
// is registered, so behavior is unchanged.
const formatters = { text: new TextFormatter() };

export function getFormatter(name) {
  const formatter = formatters[name];
  if (!formatter) throw new Error(`Unknown format: ${name}`);
  return formatter;
}

export function registerFormatter(name, formatter) {
  formatters[name] = formatter;
}

File: demo-android/app/src/main/java/com/example/reports/Formatter.kt
package com.example.reports

// A Formatter turns a report's title and rows into a string.
interface ReportFormatter {
    fun format(title: String, rows: List<Row>): String
}

class TextFormatter : ReportFormatter {
    override fun format(title: String, rows: List<Row>): String {
        val lines = mutableListOf(title, "=".repeat(title.length))
        for (row in rows) lines.add("${row.label}: ${row.value}")
        return lines.joinToString("\n")
    }
}

class CsvFormatter : ReportFormatter {
    override fun format(title: String, rows: List<Row>): String {
        val lines = mutableListOf("label,value")
        for (row in rows) lines.add("${row.label},${row.value}")
        return lines.joinToString("\n")
    }
}
